### PR TITLE
[WIP] Rework predict functions.

### PR DESCRIPTION
**Pull request by trivialfis**
Related: https://github.com/dmlc/xgboost/issues/6632  Instead of a `struct`, I went with a JSON string for any extra arguments.  Constructing `struct` in other languages are not necessarily easy.

This is a monolithic PR for early review and discussion, I will split it up into multiple smaller PRs.

## Changes

* Add new function for predicting from DMatrix.
* Use `begin_iteration` and `end_iteration` to replace `ntree_limit`.
* Change Python native, sklearn and dask interfaces for `iteration_range` and deprecate `ntree_limit`.
* Change inplace prediction functions to support `base_margin` and possibly other meta info in the future.
* Use inplace predict by default in sklearn and dask sklearn interfaces.
* Expose inplace predict in C header.
* The prediction output shape is now calculated in C++, with an additional parameter `strict_shape` to allow more explicit dimensions.  For example, when set to `True`, `predict_leaf` might output a 4 dimension array with respected to `(n_samples, n_iterations, n_classes, n_trees_in_forest)`.  Default is `False` so compatibility is kept.
* The adapters for dense and CSR are changed to accept more data types instead of strictly float32, this helps making the `inplace_predict` function to be truly inplace for all supported data types.
* All prediction function are made to be thread safe.

## Tracking extracted PRs.

- https://github.com/dmlc/xgboost/pull/6644
- https://github.com/dmlc/xgboost/pull/6645
- https://github.com/dmlc/xgboost/pull/6648
- https://github.com/dmlc/xgboost/pull/6653
- https://github.com/dmlc/xgboost/pull/6668

## Related issues:

- https://github.com/dmlc/xgboost/issues/6632
- https://github.com/dmlc/xgboost/issues/6621
- https://github.com/dmlc/xgboost/issues/6553

### CODE REVIEW

**Comment by hcho3** [comment]
>  I went with a JSON string for any extra arguments. Constructing struct in other languages are not necessarily easy.

I agree with this decision. This is also why I chose to use JSON for my global configuration PR.

File: python-package/xgboost/sklearn.py [hunk 1]
```diff
@@ -824,9 +860,18 @@ def apply(self, X, ntree_limit=0):
             ``[0; 2**(self.max_depth+1))``, possibly with gaps in the numbering.
         """
         test_dmatrix = DMatrix(X, missing=self.missing, nthread=self.n_jobs)
+        if iteration_range is None:
```
Comment (trivialfis) [inline python-package/xgboost/sklearn.py:863]: This is a breaking change, to be consistent with the `predict` function.  Not sure if it's necessary.

File: python-package/xgboost/sklearn.py [hunk 2]
```diff
@@ -1192,14 +1244,17 @@ def predict_proba(self, X, ntree_limit=None, validate_features=False,
...
+            Deprecated, use `iteration_range` instead.
         validate_features : bool
             When this is True, validate that the Booster's and data's feature_names are
             identical.  Otherwise, it is assumed that the feature_names are the same.
         base_margin : array_like
             Margin added to prediction.
+        iteration_range :
+            Specifies which layer of trees are used in prediction.  For example, if a
+            random forest is trained with 100 rounds.  Specifying `iteration_range=(10,
+            20)`, then only the forests built during [10, 20) (open set) rounds are used
```
Comment (trivialfis) [inline python-package/xgboost/sklearn.py:1256]: half open.

File: src/c_api/c_api.cc [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2014-2020 by Contributors
+// Copyright (c) 2014-2021 by Contributors
```
Comment (trivialfis) [inline src/c_api/c_api.cc:1]: I need to make more header changes.

File: src/c_api/c_api.cu [hunk 4]
```diff
@@ -30,59 +31,66 @@ XGB_DLL int XGDMatrixCreateFromArrayInterface(char const* c_json_strs,
...
-  CHECK(p_predt->DeviceCanRead());
+  CHECK(p_predt->DeviceCanRead() && !p_predt->HostCanRead());
 
+  auto &shape = learner->GetThreadLocal().prediction_shape;
+  auto chunksize = x->NumRows() == 0 ? 0 : p_predt->Size() / x->NumRows();
+  bool strict_shape = get<Boolean const>(config["strict_shape"]);
+  CalcPredictShape(strict_shape, type, x->NumRows(), x->NumColumns(), chunksize,
+                   learner->Groups(), learner->BoostedRounds(), &shape,
+                   out_dim);
+  CHECK_EQ(
```
Comment (trivialfis) [inline src/c_api/c_api.cu:64]: Move this check into calc function

File: src/common/json.cc [hunk 5]
```diff
@@ -744,4 +744,9 @@ void Json::Dump(Json json, std::string* str) {
 }
 
 Json& Json::operator=(Json const &other) = default;
+
+std::ostream &operator<<(std::ostream &os, StringView const v) {
+  os << v.c_str();
```
Comment (trivialfis) [inline src/common/json.cc:749]: Maybe a safer approach that doesn't depend on the presence of `\0`.

File: src/data/adapter.h [hunk 6]
```diff
@@ -111,60 +111,125 @@ class NoMetaInfo {
   const float* BaseMargin() const { return nullptr; }
 };
 
+inline float AdapterCast(DataType data_type, void const *values, size_t idx) {
```
Comment (trivialfis) [inline src/data/adapter.h:114]: Is there a way to merge all datatype dispatching logic?

File: src/gbm/gbtree.h [hunk 7]
```diff
@@ -260,51 +262,84 @@ class GBTree : public GradientBooster {
...
     uint32_t tree_begin, tree_end;
-    std::tie(tree_begin, tree_end) = detail::LayerToTree(model_, tparam_, layer_begin, layer_end);
-    this->GetPredictor()->InplacePredict(x, model_, missing, out_preds,
-                                         tree_begin, tree_end);
+    std::tie(tree_begin, tree_end) =
+        detail::LayerToTree(model_, tparam_, layer_begin, layer_end);
+    std::vector<Predictor const *> predictors{
+      cpu_predictor_.get(),
+#if defined(XGBOOST_USE_CUDA)
+      gpu_predictor_.get()
```
Comment (trivialfis) [inline src/gbm/gbtree.h:278]: maybe sycl?

File: src/gbm/gbtree.h [hunk 8]
```diff
@@ -260,51 +262,84 @@ class GBTree : public GradientBooster {
...
 
   void PredictLeaf(DMatrix* p_fmat,
                    HostDeviceVector<bst_float>* out_preds,
-                   unsigned ntree_limit) override {
-    this->GetPredictor()->PredictLeaf(p_fmat, out_preds, model_, ntree_limit);
+                   uint32_t layer_begin, uint32_t layer_end) override {
+    uint32_t tree_begin, tree_end;
+    std::tie(tree_begin, tree_end) = detail::LayerToTree(model_, tparam_, layer_begin, layer_end);
+    CHECK_EQ(tree_begin, 0) << "Predict leaf supports only iteration end: (0, "
+                               "n_iteration), using model slicing instead.";
```
Comment (trivialfis) [inline src/gbm/gbtree.h:314]: using -> use.

File: tests/cpp/test_learner.cc [hunk 9]
```diff
@@ -219,8 +219,13 @@ TEST(Learner, MultiThreadedPredict) {
...
       size_t constexpr kIters = 10;
       auto &entry = learner->GetThreadLocal().prediction_entry;
+      HostDeviceVector<float> predictions;
       for (size_t iter = 0; iter < kIters; ++iter) {
-        learner->Predict(p_data, false, &entry.predictions);
+        learner->Predict(p_data, false, &entry.predictions, 0, 0);
+
+        learner->Predict(p_data, false, &predictions, 0, 0, false, true);
+        learner->Predict(p_data, false, &predictions, 0, 0, false, false, true);
+        // learner->Predict(p_data, false, &predictions, 0, 0, false, false, false, false, true);
```
Comment (trivialfis) [inline tests/cpp/test_learner.cc:228]: Add it back now the size of data is dropped.

**Review by trivialfis — COMMENTED** [review]
Note to myself:

- Need more thoughts on the `strict_shape` parameter.  Should we output consistent shape when set to true?  For example the output of interaction should always be 4 dim even when classifier is not used.

**Comment by trivialfis** [comment]
This PR puts inplace prediction as the default for dask sklearn, to justify the decision, here is a profile picture for dask predict with array as input.  Not using `DaskDMatrix`, the prediction function calls `dask.array.map_blocks` and construct a small `DMatrix` for each data block.  Yellow line is the time spent on creating DMatrix.

It's possible to accelerate the prediction with `DaskDMatrix` (up to many many times), but at the cost of constructing data concatenation.

Screenshot from 2021-01-27 21-15-29
